en/guide/toolchain.md

```diff
@@ -0,0 +1,121 @@
+
+
+## Build Definition File - CMakeLists.txt
+
+*Cmake* uses a definition file named **CMakeLists.txt** to define the project. This specifies the name of the project, compiler flags for different platforms and targets, where to find dependencies (libraries and header files), source files to build, and the name of the generated binary. **CMakeLists.txt** is typically stored in the root directory of your app project.
+
+The first two lines of the file should specify the minimum compatible version of *cmake* and *your* project's name. 
+```cmake
+cmake_minimum_required(VERSION 2.8.12)
+project(your_project_name)
+```
+
+Next add the define flags that are used in compilation on different platforms 
+(you can treat this as boilerplate - it enables relatively strict handling of warnings).
+```cmake
+if(NOT MSVC)
+    add_definitions("-std=c++11 -Wall -Wextra -Werror")
+else()
+    add_definitions("-std=c++11 -WX -W2")
+    set(platform_libs "Ws2_32.lib")
+endif()
+
+# Add DEBUG define for Debug target
+set(CMAKE_CXX_FLAGS_DEBUG "-DDEBUG")
+```
+
+Add the line below to find a thread library on Linux, Mac, and Windows. This is required in order to use DroneCore!
+```cmake
+find_package(Threads REQUIRED)
+```
+
+The `include_directories()` command tells *cmake* where to find locally installed header files (if DroneCore is installed globally this is not required). 
+
+```cmake
+include_directories(
+    # Not needed if installed system-wide
+    ${CMAKE_SOURCE_DIR}/../../install/include
+)
+```
+
+> **Note** The fragment above is taken from one of the DroneCore examples, and specifies a relative path from an example sub-directory to where DroneCore exports the header files by default. 
+> If you have installed DroneCore locally and your application is not located alongside the DroneCore examples you will need to change this. 
+>
+>  Note that `CMAKE_SOURCE_DIR` is the directory in which **CMakeLists.txt** is stored, and provides a useful way of defining relative paths to the header files.
+
+Then use `add_executable()` to specify the name for your generated executable and your source files. 
+```cmake
+add_executable(your_executable_name
+    your_source_file.cpp
+)
+```
+
+The final sections of the file defines the libraries to link against. 
+The first part is used if DroneCore is installed locally, and will have to be modified depending on the relative path from your code to the **DroneCore/install/** directory.
+```cmake
+if(WINDOWS)
+    set(dronecore_lib "${CMAKE_SOURCE_DIR}/../../install/lib/dronecore.lib")
+else()
+    set(dronecore_lib "${CMAKE_SOURCE_DIR}/../../install/lib/libdronecore.a")
+endif()
+```
+
+The final section of the file defines the libraries to link against if DroneCore is installed globally.
```

Comment:
@julianoes Ditto! It is not clear to me what this does, I THINK it locates the headers using that   ${platform_libs} variable.

Reply:
No, `platform_libs` is a variable containing all libraries that you need to link to given your platform (e.g. pthread).

Reply:
Thanks. I am pretty sure based on your comments that this text and our example code are broken for use with system wide install of dronecore. I have been assuming it can magically pick up that install, but in fact nowhere do we "tell" cmake were it can find the system wide headers and binary. I don't see how it could guess except for "well known libraries"

Consider, the first bit sets the variable $dronecore_lib  to the local paths for the platform.
```cmake
if(WINDOWS)
    set(dronecore_lib "${CMAKE_SOURCE_DIR}/../../install/lib/dronecore.lib")
else()
    set(dronecore_lib "${CMAKE_SOURCE_DIR}/../../install/lib/libdronecore.a")
endif()
```
Then in this bit we link dronecore_lib variable (defined above), our platform_libs variable, and we initialise the thread library. Nowhere do we set the system libraries. The comments below are confusing - this stuff has nothing to do with system wide or dronecore.
```cmake
target_link_libraries(takeoff_and_land
    ${dronecore_lib}
    # If installed system-wide:
    # dronecore
    ${CMAKE_THREAD_LIBS_INIT}  # Link against thread library found using find_package() 
    ${platform_libs}  # Variable containing all libraries that to link against for this platform (e.g. pthread).
)
```
To work system wide cmake needs to know where system libraries are installed. 
I'd suggest that we probably should use an environment variable, but assuming hard coding, this part of the makefile should change to:
```cmake
if(WINDOWS)
    set(dronecore_lib "${CMAKE_SOURCE_DIR}/../../install/lib/dronecore.lib")
    #Using local path -not sure right thing to do.
else()
    set(dronecore_lib "/usr/local/lib/libdronecore.a")
endif()

# Bit below stays the same - ${dronecore_lib} now points to system libs. 

target_link_libraries(takeoff_and_land
    ${dronecore_lib}   #Droncore libs, as defined above
    ${CMAKE_THREAD_LIBS_INIT}  # Link against thread library found using find_package() 
    ${platform_libs}  # Variable containing all libraries that to link against for this platform (e.g. pthread).
)
```

Reply:
The thing is that the linker automatically looks at `/usr/lib/` and (for most distros) `/usr/local/lib/` for libraries. This means that if you have it installed system-wide using prefix `/usr/local`, the linking is fine with just `dronecore`.

Reply:
Ignore my comment here. You explained this well over slack :-)